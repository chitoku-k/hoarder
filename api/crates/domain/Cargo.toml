[package]
name = "domain"
authors.workspace = true
edition.workspace = true
license.workspace = true

[lib]
doctest = false

[dependencies.chrono]
version = "0.4.39"
features = ["clock", "std"]
default-features = false

[dependencies.derive_more]
version = "2.0.1"
features = ["constructor", "deref", "display", "from"]

[dependencies.futures]
version = "0.3.31"
features = ["std"]
default-features = false

[dependencies.indenter]
version = "0.3.3"

[dependencies.ordermap]
version = "0.5.5"

[dependencies.percent-encoding]
version = "2.3.1"

[dependencies.regex]
version = "1.11.1"

[dependencies.serde]
version = "1.0.218"
features = ["derive"]

[dependencies.strum]
version = "0.27.0"
features = ["derive"]

[dependencies.thiserror]
version = "2.0.11"

[dependencies.tokio]
version = "1.43.0"
features = ["macros", "rt-multi-thread"]

[dependencies.tracing]
version = "0.1.41"

[dependencies.uuid]
version = "1.13.1"
features = ["serde", "v4"]

[dev-dependencies.anyhow]
version = "1.0.95"

[dev-dependencies.indoc]
version = "2.0.5"

[dev-dependencies.mockall]
version = "0.13.1"

[dev-dependencies.pretty_assertions]
version = "1.4.1"
features = ["unstable"]

[dev-dependencies.serial_test]
version = "3.2.0"
