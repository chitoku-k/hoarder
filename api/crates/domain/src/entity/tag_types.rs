use std::hash::{Hash, Hasher};

use derive_more::{Deref, Display, From};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Clone, Copy, Debug, Default, Deref, Deserialize, Display, Eq, From, Hash, Ord, PartialEq, PartialOrd, Serialize)]
pub struct TagTypeId(Uuid);

#[derive(Clone, Debug, Eq, Ord, PartialOrd)]
pub struct TagType {
    pub id: TagTypeId,
    pub slug: String,
    pub name: String,
    pub kana: String,
}

impl Hash for TagType {
    fn hash<H>(&self, state: &mut H)
    where
        H: Hasher,
    {
        self.id.hash(state);
    }
}

impl PartialEq for TagType {
    fn eq(&self, other: &Self) -> bool {
        self.id == other.id
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashSet;

    use uuid::uuid;

    use super::*;

    #[test]
    fn tag_type_hash_equals_by_id() {
        let a = TagType {
            id: TagTypeId::from(uuid!("44444444-4444-4444-4444-444444444444")),
            slug: "foo".to_string(),
            name: "foo".to_string(),
            kana: "foo".to_string(),
        };

        let b = TagType {
            id: TagTypeId::from(uuid!("44444444-4444-4444-4444-444444444444")),
            slug: "bar".to_string(),
            name: "bar".to_string(),
            kana: "bar".to_string(),
        };

        let mut hash_set = HashSet::new();
        hash_set.insert(a);

        let actual = hash_set.contains(&b);

        assert!(actual);
    }

    #[test]
    fn tag_type_hash_not_equals_by_id() {
        let a = TagType {
            id: TagTypeId::from(uuid!("44444444-4444-4444-4444-444444444444")),
            slug: "foo".to_string(),
            name: "foo".to_string(),
            kana: "foo".to_string(),
        };

        let b = TagType {
            id: TagTypeId::from(uuid!("55555555-5555-5555-5555-555555555555")),
            slug: "foo".to_string(),
            name: "foo".to_string(),
            kana: "foo".to_string(),
        };

        let mut hash_set = HashSet::new();
        hash_set.insert(a);

        let actual = hash_set.contains(&b);

        assert!(!actual);
    }
}
