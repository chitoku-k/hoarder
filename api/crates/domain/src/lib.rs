#![allow(clippy::too_many_arguments)]

pub mod entity;
pub mod parser;
pub mod processor;
pub mod repository;
pub mod service;
