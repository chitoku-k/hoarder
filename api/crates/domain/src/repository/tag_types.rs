use std::future::Future;

use crate::{
    entity::tag_types::{TagType, TagTypeId},
    error::Result,
    repository::DeleteResult,
};

#[cfg_attr(feature = "test-mock", mockall::automock)]
pub trait TagTypesRepository: Send + Sync + 'static {
    /// Creates a tag type.
    fn create(&self, slug: &str, name: &str, kana: &str) -> impl Future<Output = Result<TagType>> + Send;

    /// Fetches all tag types.
    fn fetch_all(&self) -> impl Future<Output = Result<Vec<TagType>>> + Send;

    /// Updates the tag type by ID.
    fn update_by_id<'a>(&self, id: TagTypeId, slug: Option<&'a str>, name: Option<&'a str>, kana: Option<&'a str>) -> impl Future<Output = Result<TagType>> + Send;

    /// Deletes the tag type by Id.
    fn delete_by_id(&self, id: TagTypeId) -> impl Future<Output = Result<DeleteResult>> + Send;
}
