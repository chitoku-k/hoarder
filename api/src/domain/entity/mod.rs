pub mod external_services;
pub mod job_runs;
pub mod jobs;
pub mod media;
pub mod replicas;
pub mod sources;
pub mod tag_types;
pub mod tags;
