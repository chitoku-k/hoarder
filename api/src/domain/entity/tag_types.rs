use std::hash::{Hash, Hasher};

use derive_more::{Deref, Display, From};
use thiserror::Error;
use uuid::Uuid;

#[derive(Clone, Copy, Debug, Default, Deref, Display, Eq, From, Ord, PartialEq, PartialOrd)]
pub struct TagTypeId(Uuid);

#[derive(Clone, Debug, Eq, Ord, PartialOrd)]
pub struct TagType {
    pub id: TagTypeId,
    pub slug: String,
    pub name: String,
}

#[derive(Debug, Error)]
pub enum TagTypeError {
    #[error("tag type not found: {0}")]
    NotFound(TagTypeId),
}

impl Hash for TagType {
    fn hash<H>(&self, state: &mut H)
    where
        H: Hasher,
    {
        self.id.hash(state);
    }
}

impl PartialEq for TagType {
    fn eq(&self, other: &Self) -> bool {
        self.id == other.id
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashSet;

    use compiled_uuid::uuid;

    use super::*;

    #[test]
    fn tag_type_hash_equals_by_id() {
        let a = TagType {
            id: TagTypeId::from(uuid!("44444444-4444-4444-4444-444444444444")),
            slug: "foo".to_string(),
            name: "foo".to_string(),
        };

        let b = TagType {
            id: TagTypeId::from(uuid!("44444444-4444-4444-4444-444444444444")),
            slug: "bar".to_string(),
            name: "bar".to_string(),
        };

        let mut hash_set = HashSet::new();
        hash_set.insert(a);

        let actual = hash_set.contains(&b);

        assert!(actual);
    }

    #[test]
    fn tag_type_hash_not_equals_by_id() {
        let a = TagType {
            id: TagTypeId::from(uuid!("44444444-4444-4444-4444-444444444444")),
            slug: "foo".to_string(),
            name: "foo".to_string(),
        };

        let b = TagType {
            id: TagTypeId::from(uuid!("55555555-5555-5555-5555-555555555555")),
            slug: "foo".to_string(),
            name: "foo".to_string(),
        };

        let mut hash_set = HashSet::new();
        hash_set.insert(a);

        let actual = hash_set.contains(&b);

        assert!(!actual);
    }
}
