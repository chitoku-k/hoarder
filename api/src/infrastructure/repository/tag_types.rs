use anyhow::Context;
use async_trait::async_trait;
use derive_more::Constructor;
use futures::TryStreamExt;
use sea_query::{Expr, Iden, LockType, Order, PostgresQueryBuilder, Query};
use sea_query_binder::SqlxBinder;
use sqlx::{FromRow, PgPool};
use uuid::Uuid;

use crate::{
    domain::{
        entity::tag_types::{TagType, TagTypeError, TagTypeId},
        repository::{tag_types::TagTypesRepository, DeleteResult},
    },
    infrastructure::repository::sea_query_uuid_value,
};

#[derive(Clone, Constructor)]
pub struct PostgresTagTypesRepository {
    pool: PgPool,
}

#[derive(Debug, FromRow)]
struct PostgresTagTypeRow {
    id: Uuid,
    slug: String,
    name: String,
}

#[derive(Iden)]
pub enum PostgresTagType {
    #[iden = "tag_types"]
    Table,
    Id,
    Slug,
    Name,
}

#[derive(Iden)]
pub enum PostgresTagTagType {
    TagTypeId,
    TagTypeSlug,
    TagTypeName,
}

sea_query_uuid_value!(TagTypeId);

impl From<PostgresTagTypeRow> for TagType {
    fn from(row: PostgresTagTypeRow) -> Self {
        Self {
            id: row.id.into(),
            slug: row.slug,
            name: row.name,
        }
    }
}

#[async_trait]
impl TagTypesRepository for PostgresTagTypesRepository {
    async fn create(&self, slug: &str, name: &str) -> anyhow::Result<TagType> {
        let (sql, values) = Query::insert()
            .into_table(PostgresTagType::Table)
            .columns([PostgresTagType::Slug, PostgresTagType::Name])
            .values([Expr::val(slug).into(), Expr::val(name).into()])?
            .returning(
                Query::returning()
                    .columns([
                        PostgresTagType::Id,
                        PostgresTagType::Slug,
                        PostgresTagType::Name,
                    ])
            )
            .build_sqlx(PostgresQueryBuilder);

        let tag_type = sqlx::query_as_with::<_, PostgresTagTypeRow, _>(&sql, values)
            .fetch_one(&self.pool)
            .await?
            .into();

        Ok(tag_type)
    }

    async fn fetch_all(&self) -> anyhow::Result<Vec<TagType>> {
        let (sql, values) = Query::select()
            .columns([
                PostgresTagType::Id,
                PostgresTagType::Slug,
                PostgresTagType::Name,
            ])
            .from(PostgresTagType::Table)
            .order_by(PostgresTagType::Name, Order::Asc)
            .build_sqlx(PostgresQueryBuilder);

        let tag_types = sqlx::query_as_with::<_, PostgresTagTypeRow, _>(&sql, values)
            .fetch(&self.pool)
            .map_ok(Into::into)
            .try_collect()
            .await?;

        Ok(tag_types)
    }

    async fn update_by_id<'a>(&self, id: TagTypeId, slug: Option<&'a str>, name: Option<&'a str>) -> anyhow::Result<TagType> {
        let mut tx = self.pool.begin().await?;

        let (sql, values) = Query::select()
            .columns([
                PostgresTagType::Id,
                PostgresTagType::Slug,
                PostgresTagType::Name,
            ])
            .from(PostgresTagType::Table)
            .and_where(Expr::col(PostgresTagType::Id).eq(id))
            .lock(LockType::Update)
            .build_sqlx(PostgresQueryBuilder);

        let tag_type: TagType = sqlx::query_as_with::<_, PostgresTagTypeRow, _>(&sql, values)
            .fetch_optional(&mut tx)
            .await?
            .map(Into::into)
            .context(TagTypeError::NotFound(id))?;

        let slug = slug.unwrap_or(&tag_type.slug);
        let name = name.unwrap_or(&tag_type.name);

        let (sql, values) = Query::update()
            .table(PostgresTagType::Table)
            .value(PostgresTagType::Slug, slug)
            .value(PostgresTagType::Name, name)
            .and_where(Expr::col(PostgresTagType::Id).eq(id))
            .returning(
                Query::returning()
                    .columns([
                        PostgresTagType::Id,
                        PostgresTagType::Slug,
                        PostgresTagType::Name,
                    ])
            )
            .build_sqlx(PostgresQueryBuilder);

        let tag_type = sqlx::query_as_with::<_, PostgresTagTypeRow, _>(&sql, values)
            .fetch_one(&mut tx)
            .await?
            .into();

        tx.commit().await?;
        Ok(tag_type)
    }

    async fn delete_by_id(&self, id: TagTypeId) -> anyhow::Result<DeleteResult> {
        let (sql, values) = Query::delete()
            .from_table(PostgresTagType::Table)
            .and_where(Expr::col(PostgresTagType::Id).eq(id))
            .build_sqlx(PostgresQueryBuilder);

        let affected = sqlx::query_with(&sql, values)
            .execute(&self.pool)
            .await?
            .rows_affected();

        match affected {
            0 => Ok(DeleteResult::NotFound),
            count => Ok(DeleteResult::Deleted(count)),
        }
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;
    use sqlx::Row;
    use test_context::test_context;
    use uuid::uuid;

    use crate::infrastructure::repository::tests::DatabaseContext;

    use super::*;

    #[test_context(DatabaseContext)]
    #[tokio::test]
    #[cfg_attr(not(feature = "test-postgres"), ignore)]
    async fn create_succeeds(ctx: &DatabaseContext) {
        let repository = PostgresTagTypesRepository::new(ctx.pool.clone());
        let actual = repository.create("foobar", "FooBar").await.unwrap();

        assert_eq!(actual.slug, "foobar".to_string());
        assert_eq!(actual.name, "FooBar".to_string());

        let actual = sqlx::query(r#"SELECT "id", "slug", "name" FROM "tag_types" WHERE "id" = $1"#)
            .bind(*actual.id)
            .fetch_one(&ctx.pool)
            .await
            .unwrap();

        assert_eq!(actual.get::<&str, &str>("slug"), "foobar");
        assert_eq!(actual.get::<&str, &str>("name"), "FooBar");
    }

    #[test_context(DatabaseContext)]
    #[tokio::test]
    #[cfg_attr(not(feature = "test-postgres"), ignore)]
    async fn create_fails(ctx: &DatabaseContext) {
        let repository = PostgresTagTypesRepository::new(ctx.pool.clone());
        let actual = repository.create("character", "キャラクター").await;

        assert!(actual.is_err());
    }

    #[test_context(DatabaseContext)]
    #[tokio::test]
    #[cfg_attr(not(feature = "test-postgres"), ignore)]
    async fn fetch_all_succeeds(ctx: &DatabaseContext) {
        let repository = PostgresTagTypesRepository::new(ctx.pool.clone());
        let actual = repository.fetch_all().await.unwrap();

        assert_eq!(actual, vec![
            TagType {
                id: TagTypeId::from(uuid!("67738231-9b3a-4f45-94dc-1ba302e50e38")),
                slug: "character".to_string(),
                name: "キャラクター".to_string(),
            },
            TagType {
                id: TagTypeId::from(uuid!("1e5021f0-d8ef-4859-815a-747bf3175724")),
                slug: "work".to_string(),
                name: "作品".to_string(),
            },
            TagType {
                id: TagTypeId::from(uuid!("37553a79-53cd-4768-8a06-1378d6010954")),
                slug: "clothes".to_string(),
                name: "衣装".to_string(),
            },
        ]);
    }

    #[test_context(DatabaseContext)]
    #[tokio::test]
    #[cfg_attr(not(feature = "test-postgres"), ignore)]
    async fn update_by_id_succeeds(ctx: &DatabaseContext) {
        let repository = PostgresTagTypesRepository::new(ctx.pool.clone());
        let actual = repository.update_by_id(
            TagTypeId::from(uuid!("1e5021f0-d8ef-4859-815a-747bf3175724")),
            None,
            None,
        ).await.unwrap();

        assert_eq!(actual.slug, "work".to_string());
        assert_eq!(actual.name, "作品".to_string());

        let actual = sqlx::query(r#"SELECT "id", "slug", "name" FROM "tag_types" WHERE "id" = $1"#)
            .bind(uuid!("1e5021f0-d8ef-4859-815a-747bf3175724"))
            .fetch_one(&ctx.pool)
            .await
            .unwrap();

        assert_eq!(actual.get::<&str, &str>("slug"), "work");
        assert_eq!(actual.get::<&str, &str>("name"), "作品");
    }

    #[test_context(DatabaseContext)]
    #[tokio::test]
    #[cfg_attr(not(feature = "test-postgres"), ignore)]
    async fn update_by_id_with_slug_and_name_succeeds(ctx: &DatabaseContext) {
        let repository = PostgresTagTypesRepository::new(ctx.pool.clone());
        let actual = repository.update_by_id(
            TagTypeId::from(uuid!("1e5021f0-d8ef-4859-815a-747bf3175724")),
            Some("works"),
            Some("版権"),
        ).await.unwrap();

        assert_eq!(actual.slug, "works".to_string());
        assert_eq!(actual.name, "版権".to_string());

        let actual = sqlx::query(r#"SELECT "id", "slug", "name" FROM "tag_types" WHERE "id" = $1"#)
            .bind(uuid!("1e5021f0-d8ef-4859-815a-747bf3175724"))
            .fetch_one(&ctx.pool)
            .await
            .unwrap();

        assert_eq!(actual.get::<&str, &str>("slug"), "works");
        assert_eq!(actual.get::<&str, &str>("name"), "版権");
    }

    #[test_context(DatabaseContext)]
    #[tokio::test]
    #[cfg_attr(not(feature = "test-postgres"), ignore)]
    async fn update_by_id_fails(ctx: &DatabaseContext) {
        let repository = PostgresTagTypesRepository::new(ctx.pool.clone());
        let actual = repository.update_by_id(
            TagTypeId::from(uuid!("11111111-1111-1111-1111-111111111111")),
            Some("illustrators"),
            Some("絵師"),
        ).await;

        assert!(actual.is_err());
    }

    #[test_context(DatabaseContext)]
    #[tokio::test]
    #[cfg_attr(not(feature = "test-postgres"), ignore)]
    async fn delete_by_id_succeeds(ctx: &DatabaseContext) {
        let actual: i64 = sqlx::query(r#"SELECT COUNT(*) FROM "tag_types" WHERE "id" = $1"#)
            .bind(uuid!("1e5021f0-d8ef-4859-815a-747bf3175724"))
            .fetch_one(&ctx.pool)
            .await
            .unwrap()
            .get(0);

        assert_eq!(actual, 1);

        let repository = PostgresTagTypesRepository::new(ctx.pool.clone());
        let actual = repository.delete_by_id(TagTypeId::from(uuid!("1e5021f0-d8ef-4859-815a-747bf3175724"))).await.unwrap();

        assert_eq!(actual, DeleteResult::Deleted(1));

        let actual: i64 = sqlx::query(r#"SELECT COUNT(*) FROM "tag_types" WHERE "id" = $1"#)
            .bind(uuid!("1e5021f0-d8ef-4859-815a-747bf3175724"))
            .fetch_one(&ctx.pool)
            .await
            .unwrap()
            .get(0);

        assert_eq!(actual, 0);

        let actual = repository.delete_by_id(TagTypeId::from(uuid!("1e5021f0-d8ef-4859-815a-747bf3175724"))).await.unwrap();

        assert_eq!(actual, DeleteResult::NotFound);
    }
}
