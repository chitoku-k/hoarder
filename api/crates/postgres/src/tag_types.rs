use derive_more::{Constructor, From, Into};
use domain::{
    entity::tag_types::{TagType, TagTypeId},
    error::{Error, ErrorKind, Result},
    repository::{tag_types::TagTypesRepository, DeleteResult},
};
use futures::TryStreamExt;
use sea_query::{Expr, Iden, LockType, Order, PostgresQueryBuilder, Query};
use sea_query_binder::SqlxBinder;
use sqlx::{FromRow, PgPool};

use crate::sea_query_uuid_value;

#[derive(Clone, Constructor)]
pub struct PostgresTagTypesRepository {
    pool: PgPool,
}

#[derive(Clone, Debug, From, Into)]
pub(crate) struct PostgresTagTypeId(TagTypeId);

#[derive(Debug, FromRow)]
struct PostgresTagTypeRow {
    id: PostgresTagTypeId,
    slug: String,
    name: String,
    kana: String,
}

#[derive(Iden)]
pub(crate) enum PostgresTagType {
    #[iden = "tag_types"]
    Table,
    Id,
    Slug,
    Name,
    Kana,
}

#[derive(Iden)]
pub(crate) enum PostgresTagTagType {
    TagTypeId,
    TagTypeSlug,
    TagTypeName,
    TagTypeKana,
}

sea_query_uuid_value!(PostgresTagTypeId, TagTypeId);

impl From<PostgresTagTypeRow> for TagType {
    fn from(row: PostgresTagTypeRow) -> Self {
        Self {
            id: row.id.into(),
            slug: row.slug,
            name: row.name,
            kana: row.kana,
        }
    }
}

impl TagTypesRepository for PostgresTagTypesRepository {
    #[tracing::instrument(skip_all)]
    async fn create(&self, slug: &str, name: &str, kana: &str) -> Result<TagType> {
        let (sql, values) = Query::insert()
            .into_table(PostgresTagType::Table)
            .columns([PostgresTagType::Slug, PostgresTagType::Name, PostgresTagType::Kana])
            .values([Expr::val(slug).into(), Expr::val(name).into(), Expr::val(kana).into()])
            .map_err(Error::other)?
            .returning(
                Query::returning()
                    .columns([
                        PostgresTagType::Id,
                        PostgresTagType::Slug,
                        PostgresTagType::Name,
                        PostgresTagType::Kana,
                    ])
            )
            .build_sqlx(PostgresQueryBuilder);

        let tag_type = match sqlx::query_as_with::<_, PostgresTagTypeRow, _>(&sql, values).fetch_one(&self.pool).await {
            Ok(row) => row.into(),
            Err(sqlx::Error::Database(e)) if e.is_unique_violation() => return Err(ErrorKind::TagTypeSlugDuplicate { slug: slug.to_string() })?,
            Err(e) => return Err(Error::other(e)),
        };

        Ok(tag_type)
    }

    #[tracing::instrument(skip_all)]
    async fn fetch_by_ids<T>(&self, ids: T) -> Result<Vec<TagType>>
    where
        T: Iterator<Item = TagTypeId> + Send,
    {
        let (sql, values) = Query::select()
            .columns([
                PostgresTagType::Id,
                PostgresTagType::Slug,
                PostgresTagType::Name,
                PostgresTagType::Kana,
            ])
            .from(PostgresTagType::Table)
            .and_where(Expr::col(PostgresTagType::Id).is_in(ids.map(PostgresTagTypeId::from)))
            .order_by(PostgresTagType::Kana, Order::Asc)
            .build_sqlx(PostgresQueryBuilder);

        let tag_types = sqlx::query_as_with::<_, PostgresTagTypeRow, _>(&sql, values)
            .fetch(&self.pool)
            .map_ok(Into::into)
            .try_collect()
            .await
            .map_err(Error::other)?;

        Ok(tag_types)
    }

    #[tracing::instrument(skip_all)]
    async fn fetch_all(&self) -> Result<Vec<TagType>> {
        let (sql, values) = Query::select()
            .columns([
                PostgresTagType::Id,
                PostgresTagType::Slug,
                PostgresTagType::Name,
                PostgresTagType::Kana,
            ])
            .from(PostgresTagType::Table)
            .order_by(PostgresTagType::Kana, Order::Asc)
            .build_sqlx(PostgresQueryBuilder);

        let tag_types = sqlx::query_as_with::<_, PostgresTagTypeRow, _>(&sql, values)
            .fetch(&self.pool)
            .map_ok(Into::into)
            .try_collect()
            .await
            .map_err(Error::other)?;

        Ok(tag_types)
    }

    #[tracing::instrument(skip_all)]
    async fn update_by_id(&self, id: TagTypeId, slug: Option<&str>, name: Option<&str>, kana: Option<&str>) -> Result<TagType> {
        let mut tx = self.pool.begin().await.map_err(Error::other)?;

        let (sql, values) = Query::select()
            .columns([
                PostgresTagType::Id,
                PostgresTagType::Slug,
                PostgresTagType::Name,
                PostgresTagType::Kana,
            ])
            .from(PostgresTagType::Table)
            .and_where(Expr::col(PostgresTagType::Id).eq(PostgresTagTypeId::from(id)))
            .lock(LockType::Update)
            .build_sqlx(PostgresQueryBuilder);

        let tag_type = match sqlx::query_as_with::<_, PostgresTagTypeRow, _>(&sql, values).fetch_one(&mut *tx).await {
            Ok(row) => TagType::from(row),
            Err(sqlx::Error::RowNotFound) => return Err(ErrorKind::TagTypeNotFound { id })?,
            Err(e) => return Err(Error::other(e)),
        };

        let slug = slug.unwrap_or(&tag_type.slug);
        let name = name.unwrap_or(&tag_type.name);
        let kana = kana.unwrap_or(&tag_type.kana);

        let (sql, values) = Query::update()
            .table(PostgresTagType::Table)
            .value(PostgresTagType::Slug, slug)
            .value(PostgresTagType::Name, name)
            .value(PostgresTagType::Kana, kana)
            .and_where(Expr::col(PostgresTagType::Id).eq(PostgresTagTypeId::from(id)))
            .returning(
                Query::returning()
                    .columns([
                        PostgresTagType::Id,
                        PostgresTagType::Slug,
                        PostgresTagType::Name,
                        PostgresTagType::Kana,
                    ])
            )
            .build_sqlx(PostgresQueryBuilder);

        let tag_type = match sqlx::query_as_with::<_, PostgresTagTypeRow, _>(&sql, values).fetch_one(&mut *tx).await {
            Ok(row) => TagType::from(row),
            Err(sqlx::Error::Database(e)) if e.is_unique_violation() => return Err(ErrorKind::TagTypeSlugDuplicate { slug: slug.to_string() })?,
            Err(e) => return Err(Error::other(e)),
        };

        tx.commit().await.map_err(Error::other)?;
        Ok(tag_type)
    }

    #[tracing::instrument(skip_all)]
    async fn delete_by_id(&self, id: TagTypeId) -> Result<DeleteResult> {
        let (sql, values) = Query::delete()
            .from_table(PostgresTagType::Table)
            .and_where(Expr::col(PostgresTagType::Id).eq(PostgresTagTypeId::from(id)))
            .build_sqlx(PostgresQueryBuilder);

        let affected = sqlx::query_with(&sql, values)
            .execute(&self.pool)
            .await
            .map_err(Error::other)?
            .rows_affected();

        match affected {
            0 => Ok(DeleteResult::NotFound),
            count => Ok(DeleteResult::Deleted(count)),
        }
    }
}
