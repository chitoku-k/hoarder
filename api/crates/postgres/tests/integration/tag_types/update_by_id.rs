use domain::{
    entity::tag_types::TagTypeId,
    error::ErrorKind,
    repository::tag_types::TagTypesRepository,
};
use insta::assert_toml_snapshot;
use postgres::tag_types::PostgresTagTypesRepository;
use pretty_assertions::{assert_eq, assert_matches};
use sqlx::Row;
use test_context::test_context;
use tracing::Instrument;
use uuid::uuid;

use super::DatabaseContext;

#[test_context(DatabaseContext)]
#[tokio::test]
async fn succeeds(ctx: &DatabaseContext) {
    let repository = PostgresTagTypesRepository::new(ctx.pool.clone());
    let actual = repository.update_by_id(
        TagTypeId::from(uuid!("1e5021f0-d8ef-4859-815a-747bf3175724")),
        None,
        None,
        None,
    ).instrument(ctx.span.clone()).await.unwrap();

    assert_eq!(actual.slug, "work".to_string());
    assert_eq!(actual.name, "作品".to_string());
    assert_eq!(actual.kana, "さくひん".to_string());

    let actual = sqlx::query(r#"SELECT "id", "slug", "name", "kana" FROM "tag_types" WHERE "id" = $1"#)
        .bind(uuid!("1e5021f0-d8ef-4859-815a-747bf3175724"))
        .fetch_one(&ctx.pool)
        .await
        .unwrap();

    assert_eq!(actual.get::<&str, &str>("slug"), "work");
    assert_eq!(actual.get::<&str, &str>("name"), "作品");
    assert_eq!(actual.get::<&str, &str>("kana"), "さくひん");

    assert_toml_snapshot!(ctx.queries());
}

#[test_context(DatabaseContext)]
#[tokio::test]
async fn with_slug_name_kana_succeeds(ctx: &DatabaseContext) {
    let repository = PostgresTagTypesRepository::new(ctx.pool.clone());
    let actual = repository.update_by_id(
        TagTypeId::from(uuid!("1e5021f0-d8ef-4859-815a-747bf3175724")),
        Some("works"),
        Some("版権"),
        Some("はんけん"),
    ).instrument(ctx.span.clone()).await.unwrap();

    assert_eq!(actual.slug, "works".to_string());
    assert_eq!(actual.name, "版権".to_string());
    assert_eq!(actual.kana, "はんけん".to_string());

    let actual = sqlx::query(r#"SELECT "id", "slug", "name", "kana" FROM "tag_types" WHERE "id" = $1"#)
        .bind(uuid!("1e5021f0-d8ef-4859-815a-747bf3175724"))
        .fetch_one(&ctx.pool)
        .await
        .unwrap();

    assert_eq!(actual.get::<&str, &str>("slug"), "works");
    assert_eq!(actual.get::<&str, &str>("name"), "版権");
    assert_eq!(actual.get::<&str, &str>("kana"), "はんけん");

    assert_toml_snapshot!(ctx.queries());
}

#[test_context(DatabaseContext)]
#[tokio::test]
async fn fails(ctx: &DatabaseContext) {
    let repository = PostgresTagTypesRepository::new(ctx.pool.clone());
    let actual = repository.update_by_id(
        TagTypeId::from(uuid!("11111111-1111-1111-1111-111111111111")),
        Some("illustrators"),
        Some("絵師"),
        Some("えし"),
    ).instrument(ctx.span.clone()).await.unwrap_err();

    assert_matches!(actual.kind(), ErrorKind::TagTypeNotFound { id } if id == &TagTypeId::from(uuid!("11111111-1111-1111-1111-111111111111")));

    assert_toml_snapshot!(ctx.queries());
}
