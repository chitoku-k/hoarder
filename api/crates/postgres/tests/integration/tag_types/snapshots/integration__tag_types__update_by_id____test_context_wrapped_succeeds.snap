---
source: crates/postgres/tests/integration/tag_types/update_by_id.rs
expression: ctx.queries()
---
[[queries]]
sql = '''
SELECT
  "id",
  "slug",
  "name",
  "kana"
FROM
  "tag_types"
WHERE
  "id" = $1
FOR UPDATE'''
rows_affected = 1
rows_returned = 1

[[queries]]
sql = '''
UPDATE
  "tag_types"
SET
  "slug" = $1,
  "name" = $2,
  "kana" = $3
WHERE
  "id" = $4
RETURNING
  "id",
  "slug",
  "name",
  "kana"'''
rows_affected = 1
rows_returned = 1

[[queries]]
sql = 'COMMIT'
rows_affected = 0
rows_returned = 0
